src/Analyzer.Utilities/JTokenExtensions.cs

```diff
@@ -13,107 +15,320 @@ namespace Microsoft.Azure.Templates.Analyzer.Utilities
+            var jtokensToSearchFrom = new Queue<InsensitiveTokenContext>();
+            jtokensToSearchFrom.Enqueue(new InsensitiveTokenContext { Token = token, ResolvedPath = new List<string>(), NextPathIndex = 0 });
+
+            bool fullPathFound = false;
+
+            while (jtokensToSearchFrom.Count > 0)
             {
-                pathSoFar.Add(propertyName);
+                var tokenContext = jtokensToSearchFrom.Dequeue();
+
+                if (tokenContext.NextPathIndex == properties.Length)
+                {
+                    // Reached the end of the path.  Return current JToken.
+                    yield return tokenContext.Token;
+                    fullPathFound = true;
+                    continue;
+                }
+
+                var propertyToFind = properties[tokenContext.NextPathIndex];
+
+                if (string.Equals(propertyToFind, "*"))
+                {
+                    // Wildcard - add all children.
+                    AddAllChildrenToQueue(jtokensToSearchFrom, tokenContext);
+                    continue;
+                }
+
+                // See if property in requested path is indexing an array
+                var arrayMatch = jArrayRegex.Match(propertyToFind);
+                if (arrayMatch.Success)
+                {
+                    propertyToFind = arrayMatch.Groups["property"].Value;
+                }
+
                 bool foundChild = false;
                 Exception lastException = null;
-                var arrayMatch = jArrayRegex.Match(propertyName);
-
-                foreach (var child in token.Children())
+                foreach (var child in tokenContext.Token.Children())
                 {
                     if (child is JProperty childProperty)
                     {
-                        if (arrayMatch.Success && string.Equals(arrayMatch.Groups["property"].Value, childProperty.Name, StringComparison.OrdinalIgnoreCase))
+                        if (string.Equals(propertyToFind, childProperty.Name, StringComparison.OrdinalIgnoreCase))
                         {
-                            try
+                            // Found the target property.
+                            foundChild = true;
+                            if (arrayMatch.Success)
                             {
-                                var index = int.Parse(arrayMatch.Groups["index"].Value);
-                                token = childProperty.Value[index];
-                                foundChild = true;
+                                // Path is indexing into this array property.
+                                // Attempt to find the requested index.
+                                if (childProperty.Value is JArray childArray)
+                                {
+                                    var index = arrayMatch.Groups["index"].Value;
+                                    if (index == "*")
+                                    {
```

Comment:
Could be better to put line 197 foundChild = true; in the if condition instead here as you are just declaring the bool in line 188, then later putting true, then putting false if the condition breaks.

Reply:
It looks like no matter how I do it (setting `true` only when absolutely sure it's true, or setting `false` if a condition breaks), I have to explicitly set `foundChild` to something in three separate places (either three `true`s or three `false`s).  So I guess it's just preference.  Unless you're feeling strongly one way, I think I'll leave it as-is.